<center><H1><b> DS 203 Assignment 7 </b></H1></center>
<center><H2><b> Name: Vinit Awale </b></H2></center>
<center><H2><b> Roll No: 18D070067 </b></H2></center>

## QUESTION 3

### IMPORTS

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

### READING THE DATASET


In [2]:
df = pd.read_csv(r"Datasets\archive\summer-products-with-rating-and-performance_2020-08.csv")

In [3]:
df.head()

title  \
0  2020 Summer Vintage Flamingo Print  Pajamas Se...   
1  SSHOUSE Summer Casual Sleeveless Soirée Party ...   
2  2020 Nouvelle Arrivée Femmes Printemps et Été ...   
3  Hot Summer Cool T-shirt pour les femmes Mode T...   
4  Femmes Shorts d'été à lacets taille élastique ...   

                                          title_orig  price  retail_price  \
0  2020 Summer Vintage Flamingo Print  Pajamas Se...  16.00            14   
1   Women's Casual Summer Sleeveless Sexy Mini Dress   8.00            22   
2  2020 New Arrival Women Spring and Summer Beach...   8.00            43   
3  Hot Summer Cool T Shirt for Women Fashion Tops...   8.00             8   
4  Women Summer Shorts Lace Up Elastic Waistband ...   2.72             3   

  currency_buyer  units_sold  uses_ad_boosts  rating  rating_count  \
0            EUR         100               0    3.76            54   
1            EUR       20000               1    3.45          6135   
2            EUR         100               0    3.57            14   
3            EUR        5000               1    4.03           579   
4            EUR         100               1    3.10            20   

   rating_five_count  ...  merchant_rating_count  merchant_rating  \
0               26.0  ...                    568         4.128521   
1             2269.0  ...                  17752         3.899673   
2                5.0  ...                    295         3.989831   
3              295.0  ...                  23832         4.020435   
4                6.0  ...                  14482         4.001588   

                merchant_id  merchant_has_profile_picture  \
0  595097d6a26f6e070cb878d1                             0   
1  56458aa03a698c35c9050988                             0   
2  5d464a1ffdf7bc44ee933c65                             0   
3  58cfdefdacb37b556efdff7c                             0   
4  5ab3b592c3911a095ad5dadb                             0   

   merchant_profile_picture                                      product_url  \
0                       NaN  https://www.wish.com/c/5e9ae51d43d6a96e303acdb0   
1                       NaN  https://www.wish.com/c/58940d436a0d3d5da4e95a38   
2                       NaN  https://www.wish.com/c/5ea10e2c617580260d55310a   
3                       NaN  https://www.wish.com/c/5cedf17ad1d44c52c59e4aca   
4                       NaN  https://www.wish.com/c/5ebf5819ebac372b070b0e70   

                                     product_picture  \
0  https://contestimg.wish.com/api/webimage/5e9ae...   
1  https://contestimg.wish.com/api/webimage/58940...   
2  https://contestimg.wish.com/api/webimage/5ea10...   
3  https://contestimg.wish.com/api/webimage/5cedf...   
4  https://contestimg.wish.com/api/webimage/5ebf5...   

                 product_id   theme crawl_month  
0  5e9ae51d43d6a96e303acdb0  summer     2020-08  
1  58940d436a0d3d5da4e95a38  summer     2020-08  
2  5ea10e2c617580260d55310a  summer     2020-08  
3  5cedf17ad1d44c52c59e4aca  summer     2020-08  
4  5ebf5819ebac372b070b0e70  summer     2020-08  

[5 rows x 43 columns]

### Answer a) 
 - From the dataset we can easily see that the variable to be predicted i.e. the rating is a continuous variable. 
 - Also, clearly this is a labeled dataset, i.e. we know the rating of the cloth items. Hence, we are solving a supervised learning problem.
 - The target variable is the rating of the cloth items, given by 'rating' column in the dataset. 

### Answer b) # TODO

### Answer c) Let us look at all the features in the dataset


In [4]:
df.columns

Index(['title', 'title_orig', 'price', 'retail_price', 'currency_buyer',
       'units_sold', 'uses_ad_boosts', 'rating', 'rating_count',
       'rating_five_count', 'rating_four_count', 'rating_three_count',
       'rating_two_count', 'rating_one_count', 'badges_count',
       'badge_local_product', 'badge_product_quality', 'badge_fast_shipping',
       'tags', 'product_color', 'product_variation_size_id',
       'product_variation_inventory', 'shipping_option_name',
       'shipping_option_price', 'shipping_is_express', 'countries_shipped_to',
       'inventory_total', 'has_urgency_banner', 'urgency_text',
       'origin_country', 'merchant_title', 'merchant_name',
       'merchant_info_subtitle', 'merchant_rating_count', 'merchant_rating',
       'merchant_id', 'merchant_has_profile_picture',
       'merchant_profile_picture', 'product_url', 'product_picture',
       'product_id', 'theme', 'crawl_month'],
      dtype='object')

The variables which I feel will be highly predictive of the rating of the cloth items are:  
- 'price'  
- 'retail_price'  
- 'units_sold'  
- 'badges_count'  
- 'badge_local_product'  
- 'badge_product_quality'  
- 'badge_fast_shipping'  
- 'merchant_rating_count'  
- 'merchant_rating'  
- 'theme'

Also, the following variables are not available at the test time and hence will not be used in the model:
- 'rating_five_count'
- 'rating_four_count'
- 'rating_three_count'
- 'rating_two_count'
- 'rating_one_count'

This is because if we have these features beforehand, a simple weighted average of these features will give us the rating of the cloth item. Hence, we remove these features from the dataset.

In [5]:
features_remove = ['rating_five_count' , 'rating_four_count', 'rating_three_count', 'rating_two_count' , 'rating_one_count']

df = df.drop(features_remove, axis = 1)

In [6]:
df.head()

title  \
0  2020 Summer Vintage Flamingo Print  Pajamas Se...   
1  SSHOUSE Summer Casual Sleeveless Soirée Party ...   
2  2020 Nouvelle Arrivée Femmes Printemps et Été ...   
3  Hot Summer Cool T-shirt pour les femmes Mode T...   
4  Femmes Shorts d'été à lacets taille élastique ...   

                                          title_orig  price  retail_price  \
0  2020 Summer Vintage Flamingo Print  Pajamas Se...  16.00            14   
1   Women's Casual Summer Sleeveless Sexy Mini Dress   8.00            22   
2  2020 New Arrival Women Spring and Summer Beach...   8.00            43   
3  Hot Summer Cool T Shirt for Women Fashion Tops...   8.00             8   
4  Women Summer Shorts Lace Up Elastic Waistband ...   2.72             3   

  currency_buyer  units_sold  uses_ad_boosts  rating  rating_count  \
0            EUR         100               0    3.76            54   
1            EUR       20000               1    3.45          6135   
2            EUR         100               0    3.57            14   
3            EUR        5000               1    4.03           579   
4            EUR         100               1    3.10            20   

   badges_count  ...  merchant_rating_count  merchant_rating  \
0             0  ...                    568         4.128521   
1             0  ...                  17752         3.899673   
2             0  ...                    295         3.989831   
3             0  ...                  23832         4.020435   
4             0  ...                  14482         4.001588   

                merchant_id merchant_has_profile_picture  \
0  595097d6a26f6e070cb878d1                            0   
1  56458aa03a698c35c9050988                            0   
2  5d464a1ffdf7bc44ee933c65                            0   
3  58cfdefdacb37b556efdff7c                            0   
4  5ab3b592c3911a095ad5dadb                            0   

  merchant_profile_picture                                      product_url  \
0                      NaN  https://www.wish.com/c/5e9ae51d43d6a96e303acdb0   
1                      NaN  https://www.wish.com/c/58940d436a0d3d5da4e95a38   
2                      NaN  https://www.wish.com/c/5ea10e2c617580260d55310a   
3                      NaN  https://www.wish.com/c/5cedf17ad1d44c52c59e4aca   
4                      NaN  https://www.wish.com/c/5ebf5819ebac372b070b0e70   

                                     product_picture  \
0  https://contestimg.wish.com/api/webimage/5e9ae...   
1  https://contestimg.wish.com/api/webimage/58940...   
2  https://contestimg.wish.com/api/webimage/5ea10...   
3  https://contestimg.wish.com/api/webimage/5cedf...   
4  https://contestimg.wish.com/api/webimage/5ebf5...   

                 product_id   theme  crawl_month  
0  5e9ae51d43d6a96e303acdb0  summer      2020-08  
1  58940d436a0d3d5da4e95a38  summer      2020-08  
2  5ea10e2c617580260d55310a  summer      2020-08  
3  5cedf17ad1d44c52c59e4aca  summer      2020-08  
4  5ebf5819ebac372b070b0e70  summer      2020-08  

[5 rows x 38 columns]

From these features the ones which will not contribute to the rating of the cloth item are:
- title and title_orig (Since this is different for all the items and hence nothing can be generalised from it)
- 'shipping_option_name'
- 'shipping_option_price'
- 'shipping_is_express'
- 'countries_shipped_to'
- 'inventory_total'
- 'has_urgency_banner'
- 'urgency_text'

However, we will not simply reject these features as they may be useful in the future.

We will drop the following features since they are simply the unqiue identifier of the cloth item:
- title
- title_orig
- product_id

In [7]:
features_remove = ['title', 'title_orig', 'product_id']

df = df.drop(features_remove, axis = 1)

In [8]:
df.head()

price  retail_price currency_buyer  units_sold  uses_ad_boosts  rating  \
0  16.00            14            EUR         100               0    3.76   
1   8.00            22            EUR       20000               1    3.45   
2   8.00            43            EUR         100               0    3.57   
3   8.00             8            EUR        5000               1    4.03   
4   2.72             3            EUR         100               1    3.10   

   rating_count  badges_count  badge_local_product  badge_product_quality  \
0            54             0                    0                      0   
1          6135             0                    0                      0   
2            14             0                    0                      0   
3           579             0                    0                      0   
4            20             0                    0                      0   

   ...             merchant_info_subtitle merchant_rating_count  \
0  ...                        (568 notes)                   568   
1  ...  83 % avis positifs (17,752 notes)                 17752   
2  ...     86 % avis positifs (295 notes)                   295   
3  ...                     (23,832 notes)                 23832   
4  ...  85 % avis positifs (14,482 notes)                 14482   

  merchant_rating               merchant_id  merchant_has_profile_picture  \
0        4.128521  595097d6a26f6e070cb878d1                             0   
1        3.899673  56458aa03a698c35c9050988                             0   
2        3.989831  5d464a1ffdf7bc44ee933c65                             0   
3        4.020435  58cfdefdacb37b556efdff7c                             0   
4        4.001588  5ab3b592c3911a095ad5dadb                             0   

  merchant_profile_picture                                      product_url  \
0                      NaN  https://www.wish.com/c/5e9ae51d43d6a96e303acdb0   
1                      NaN  https://www.wish.com/c/58940d436a0d3d5da4e95a38   
2                      NaN  https://www.wish.com/c/5ea10e2c617580260d55310a   
3                      NaN  https://www.wish.com/c/5cedf17ad1d44c52c59e4aca   
4                      NaN  https://www.wish.com/c/5ebf5819ebac372b070b0e70   

                                     product_picture   theme  crawl_month  
0  https://contestimg.wish.com/api/webimage/5e9ae...  summer      2020-08  
1  https://contestimg.wish.com/api/webimage/58940...  summer      2020-08  
2  https://contestimg.wish.com/api/webimage/5ea10...  summer      2020-08  
3  https://contestimg.wish.com/api/webimage/5cedf...  summer      2020-08  
4  https://contestimg.wish.com/api/webimage/5ebf5...  summer      2020-08  

[5 rows x 35 columns]

### Answer d)
All the features in the dataframe are useable expect these:
- 'merchant_profile_picture'
- 'product_url'
- 'product_picture'


Hence we drop these features from the dataset.

In [9]:
features_remove = ['merchant_profile_picture', 'product_url', 'product_picture']

df = df.drop(features_remove, axis = 1)

## HANDLING MISSING VALUES

In [10]:
df.isna().sum()

price                              0
retail_price                       0
currency_buyer                     0
units_sold                         0
uses_ad_boosts                     0
rating                             0
rating_count                       0
badges_count                       0
badge_local_product                0
badge_product_quality              0
badge_fast_shipping                0
tags                               0
product_color                     41
product_variation_size_id         14
product_variation_inventory        0
shipping_option_name               0
shipping_option_price              0
shipping_is_express                0
countries_shipped_to               0
inventory_total                    0
has_urgency_banner              1100
urgency_text                    1100
origin_country                    17
merchant_title                     0
merchant_name                      4
merchant_info_subtitle             1
merchant_rating_count              0
m

The features with missing values are:
- product_color
- product_variation_size_id 
- has_urgency_banner
- urgency_text
- origin_country
- merchant_name 
- merchant_info_subtitle

In [11]:
features_remove = ['product_color',
'product_variation_size_id', 
'has_urgency_banner',
'urgency_text',
'origin_country',
'merchant_name', 
'merchant_info_subtitle']

df = df.drop(features_remove, axis = 1)

In [12]:
df.head()

price  retail_price currency_buyer  units_sold  uses_ad_boosts  rating  \
0  16.00            14            EUR         100               0    3.76   
1   8.00            22            EUR       20000               1    3.45   
2   8.00            43            EUR         100               0    3.57   
3   8.00             8            EUR        5000               1    4.03   
4   2.72             3            EUR         100               1    3.10   

   rating_count  badges_count  badge_local_product  badge_product_quality  \
0            54             0                    0                      0   
1          6135             0                    0                      0   
2            14             0                    0                      0   
3           579             0                    0                      0   
4            20             0                    0                      0   

   ...  shipping_is_express countries_shipped_to  inventory_total  \
0  ...                    0                   34               50   
1  ...                    0                   41               50   
2  ...                    0                   36               50   
3  ...                    0                   41               50   
4  ...                    0                   35               50   

    merchant_title  merchant_rating_count  merchant_rating  \
0         zgrdejia                    568         4.128521   
1        SaraHouse                  17752         3.899673   
2           hxt520                    295         3.989831   
3         allenfan                  23832         4.020435   
4  youngpeopleshop                  14482         4.001588   

                merchant_id  merchant_has_profile_picture   theme  crawl_month  
0  595097d6a26f6e070cb878d1                             0  summer      2020-08  
1  56458aa03a698c35c9050988                             0  summer      2020-08  
2  5d464a1ffdf7bc44ee933c65                             0  summer      2020-08  
3  58cfdefdacb37b556efdff7c                             0  summer      2020-08  
4  5ab3b592c3911a095ad5dadb                             0  summer      2020-08  

[5 rows x 25 columns]

In [13]:
# Get unique values for each object feature
for feature in df.columns:
    print(feature, df[feature].nunique())


price 127
retail_price 104
currency_buyer 1
units_sold 15
uses_ad_boosts 2
rating 192
rating_count 761
badges_count 4
badge_local_product 2
badge_product_quality 2
badge_fast_shipping 2
tags 1230
product_variation_inventory 48
shipping_option_name 15
shipping_option_price 8
shipping_is_express 2
countries_shipped_to 94
inventory_total 10
merchant_title 958
merchant_rating_count 917
merchant_rating 952
merchant_id 958
merchant_has_profile_picture 2
theme 1
crawl_month 1


We reject categorical features with more than 50% unique values. Also, we reject features with only one unique value.

In [14]:
df.shape

(1573, 25)

Since we have 1573 entries in the dataset, we reject the categorical features with more than 50% unique values i.e. we reject the features with more than 786 unique values. Also, we reject the features with only one unique value. Hence we drop the following features from the dataset:
- merchant_title
- merchant_id
- theme
- crawl_month
- currency_buyer
- tags

In [15]:
features_remove = ['merchant_title' , 'merchant_id' , 'theme', 'crawl_month', 'currency_buyer', 'tags']

df = df.drop(features_remove, axis = 1)

In [16]:
df.dtypes

price                           float64
retail_price                      int64
units_sold                        int64
uses_ad_boosts                    int64
rating                          float64
rating_count                      int64
badges_count                      int64
badge_local_product               int64
badge_product_quality             int64
badge_fast_shipping               int64
product_variation_inventory       int64
shipping_option_name             object
shipping_option_price             int64
shipping_is_express               int64
countries_shipped_to              int64
inventory_total                   int64
merchant_rating_count             int64
merchant_rating                 float64
merchant_has_profile_picture      int64
dtype: object

In [17]:
# Features with dtype object
object_features = df.select_dtypes(include=['object']).columns

In [18]:
object_features

Index(['shipping_option_name'], dtype='object')

### ONE HOT ENCODING

In [19]:
# One hot encoding for object_features
df = pd.get_dummies(df, columns=object_features)

In [20]:
df.head()

price  retail_price  units_sold  uses_ad_boosts  rating  rating_count  \
0  16.00            14         100               0    3.76            54   
1   8.00            22       20000               1    3.45          6135   
2   8.00            43         100               0    3.57            14   
3   8.00             8        5000               1    4.03           579   
4   2.72             3         100               1    3.10            20   

   badges_count  badge_local_product  badge_product_quality  \
0             0                    0                      0   
1             0                    0                      0   
2             0                    0                      0   
3             0                    0                      0   
4             0                    0                      0   

   badge_fast_shipping  ...  shipping_option_name_Livraison standard  \
0                    0  ...                                        1   
1                    0  ...                                        1   
2                    0  ...                                        1   
3                    0  ...                                        1   
4                    0  ...                                        1   

   shipping_option_name_Spedizione standard  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   

   shipping_option_name_Standard Shipping  \
0                                       0   
1                                       0   
2                                       0   
3                                       0   
4                                       0   

   shipping_option_name_Standardowa wysyłka  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   

   shipping_option_name_Standardversand  \
0                                     0   
1                                     0   
2                                     0   
3                                     0   
4                                     0   

   shipping_option_name_Standart Gönderi  \
0                                      0   
1                                      0   
2                                      0   
3                                      0   
4                                      0   

   shipping_option_name_Стандартная доставка  \
0                                          0   
1                                          0   
2                                          0   
3                                          0   
4                                          0   

   shipping_option_name_الشحن القياسي  \
0                                   0   
1                                   0   
2                                   0   
3                                   0   
4                                   0   

   shipping_option_name_การส่งสินค้ามาตรฐาน  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   

   shipping_option_name_ការដឹកជញ្ជូនតាមស្តង់ដារ  
0                                             0  
1                                             0  
2                                             0  
3                                             0  
4                                             0  

[5 rows x 33 columns]

In [21]:
df.dtypes

price                                           float64
retail_price                                      int64
units_sold                                        int64
uses_ad_boosts                                    int64
rating                                          float64
rating_count                                      int64
badges_count                                      int64
badge_local_product                               int64
badge_product_quality                             int64
badge_fast_shipping                               int64
product_variation_inventory                       int64
shipping_option_price                             int64
shipping_is_express                               int64
countries_shipped_to                              int64
inventory_total                                   int64
merchant_rating_count                             int64
merchant_rating                                 float64
merchant_has_profile_picture                    

### Normalization of features with dtype 'int64' and 'float64'

In [22]:
# Find features with dtpe int64 and float64

numerical_features = df.select_dtypes(include=['int64' , 'float64']).columns

In [23]:
numerical_features

Index(['price', 'retail_price', 'units_sold', 'uses_ad_boosts', 'rating',
       'rating_count', 'badges_count', 'badge_local_product',
       'badge_product_quality', 'badge_fast_shipping',
       'product_variation_inventory', 'shipping_option_price',
       'shipping_is_express', 'countries_shipped_to', 'inventory_total',
       'merchant_rating_count', 'merchant_rating',
       'merchant_has_profile_picture'],
      dtype='object')

Of these features the features 'merchant_has_profile_picture' and 'shipping_is_express' are categorical features. Hence we will not normalize these features.

In [24]:
numerical_features = numerical_features.drop(['merchant_has_profile_picture' ,'shipping_is_express'])

In [25]:
# Normalize numerical features

for feature in numerical_features:
    df[feature] = (df[feature] - np.mean(df[feature])) / np.std(df[feature])

In [26]:
df.head()

price  retail_price  units_sold  uses_ad_boosts    rating  rating_count  \
0  1.952444     -0.306068   -0.453197       -0.873758 -0.118197     -0.421348   
1 -0.082775     -0.042461    1.674334        1.144482 -0.719894      2.644757   
2 -0.082775      0.649507   -0.453197       -0.873758 -0.486979     -0.441517   
3 -0.082775     -0.503773    0.070668        1.144482  0.405861     -0.156638   
4 -1.426020     -0.668528   -0.453197        1.144482 -1.399229     -0.438491   

   badges_count  badge_local_product  badge_product_quality  \
0     -0.309837            -0.137049              -0.283473   
1     -0.309837            -0.137049              -0.283473   
2     -0.309837            -0.137049              -0.283473   
3     -0.309837            -0.137049              -0.283473   
4     -0.309837            -0.137049              -0.283473   

   badge_fast_shipping  ...  shipping_option_name_Livraison standard  \
0            -0.113483  ...                                        1   
1            -0.113483  ...                                        1   
2            -0.113483  ...                                        1   
3            -0.113483  ...                                        1   
4            -0.113483  ...                                        1   

   shipping_option_name_Spedizione standard  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   

   shipping_option_name_Standard Shipping  \
0                                       0   
1                                       0   
2                                       0   
3                                       0   
4                                       0   

   shipping_option_name_Standardowa wysyłka  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   

   shipping_option_name_Standardversand  \
0                                     0   
1                                     0   
2                                     0   
3                                     0   
4                                     0   

   shipping_option_name_Standart Gönderi  \
0                                      0   
1                                      0   
2                                      0   
3                                      0   
4                                      0   

   shipping_option_name_Стандартная доставка  \
0                                          0   
1                                          0   
2                                          0   
3                                          0   
4                                          0   

   shipping_option_name_الشحن القياسي  \
0                                   0   
1                                   0   
2                                   0   
3                                   0   
4                                   0   

   shipping_option_name_การส่งสินค้ามาตรฐาน  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   

   shipping_option_name_ការដឹកជញ្ជូនតាមស្តង់ដារ  
0                                             0  
1                                             0  
2                                             0  
3                                             0  
4                                             0  

[5 rows x 33 columns]

In [27]:
df.shape

(1573, 33)

We can see that the df has 33 features. Hence, we need to reduce the dimensionality of the dataset.

In [28]:
# Split the dataset in X and y
X = df.drop('rating', axis = 1)
y = df['rating']
numerical_features = numerical_features.drop(['rating'])
object_features = df.columns.difference(numerical_features)
object_features = object_features.drop('rating')

In [29]:
object_features

Index(['merchant_has_profile_picture', 'shipping_is_express',
       'shipping_option_name_Ekspresowa wysyłka',
       'shipping_option_name_Envio Padrão',
       'shipping_option_name_Envío normal',
       'shipping_option_name_Expediere Standard',
       'shipping_option_name_Livraison Express',
       'shipping_option_name_Livraison standard',
       'shipping_option_name_Spedizione standard',
       'shipping_option_name_Standard Shipping',
       'shipping_option_name_Standardowa wysyłka',
       'shipping_option_name_Standardversand',
       'shipping_option_name_Standart Gönderi',
       'shipping_option_name_Стандартная доставка',
       'shipping_option_name_الشحن القياسي',
       'shipping_option_name_การส่งสินค้ามาตรฐาน',
       'shipping_option_name_ការដឹកជញ្ជូនតាមស្តង់ដារ'],
      dtype='object')

In [30]:
numerical_features.nunique()

15

Hence there are 15 numerical features in total. By performing PCA we bring that down to 5 features.

### Answer f)

In [31]:
# PCA for the features in the dataset which are numerical
from sklearn.decomposition import PCA


pca = PCA(n_components=5)

X_pca = pca.fit_transform(X[numerical_features])

In [32]:
X_pca

array([[ 0.86419104,  2.03205615, -1.35852029,  0.32611946, -0.3939176 ],
       [ 0.42801654, -2.14051137, -1.58303345, -1.26324632,  0.01886661],
       [-0.59905336,  0.9360163 ,  0.07900654, -0.46690491, -0.23414291],
       ...,
       [-1.26875779,  0.17017511,  0.58444635, -0.63822214, -1.0225359 ],
       [-0.31280568,  1.50445946, -0.59573822, -0.43959098, -0.83470727],
       [-0.50492062,  0.0201617 ,  0.30190665,  0.61353769,  0.53597607]])

In [33]:
# Concatanate the PCA features with the categorical features
X_pca = pd.DataFrame(X_pca)
X_pca.columns = ['PCA_' + str(i) for i in range(1,6)]

# Concatenate the dataframes
X_concat = pd.concat([X[object_features], X_pca], axis = 1)


In [34]:
X_concat.head()

merchant_has_profile_picture  shipping_is_express  \
0                             0                    0   
1                             0                    0   
2                             0                    0   
3                             0                    0   
4                             0                    0   

   shipping_option_name_Ekspresowa wysyłka  shipping_option_name_Envio Padrão  \
0                                        0                                  0   
1                                        0                                  0   
2                                        0                                  0   
3                                        0                                  0   
4                                        0                                  0   

   shipping_option_name_Envío normal  shipping_option_name_Expediere Standard  \
0                                  0                                        0   
1                                  0                                        0   
2                                  0                                        0   
3                                  0                                        0   
4                                  0                                        0   

   shipping_option_name_Livraison Express  \
0                                       0   
1                                       0   
2                                       0   
3                                       0   
4                                       0   

   shipping_option_name_Livraison standard  \
0                                        1   
1                                        1   
2                                        1   
3                                        1   
4                                        1   

   shipping_option_name_Spedizione standard  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   

   shipping_option_name_Standard Shipping  ...  \
0                                       0  ...   
1                                       0  ...   
2                                       0  ...   
3                                       0  ...   
4                                       0  ...   

   shipping_option_name_Standart Gönderi  \
0                                      0   
1                                      0   
2                                      0   
3                                      0   
4                                      0   

   shipping_option_name_Стандартная доставка  \
0                                          0   
1                                          0   
2                                          0   
3                                          0   
4                                          0   

   shipping_option_name_الشحن القياسي  \
0                                   0   
1                                   0   
2                                   0   
3                                   0   
4                                   0   

   shipping_option_name_การส่งสินค้ามาตรฐาน  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   

   shipping_option_name_ការដឹកជញ្ជូនតាមស្តង់ដារ     PCA_1     PCA_2     PCA_3  \
0                                             0  0.864191  2.032056 -1.358520   
1                                             0  0.428017 -2.140511 -1.583033   
2                                             0 -0.599053  0.936016  0.079007   
3                                             0 -0.470319 -0.185161 -0.084137   
4                                             0 -1.959373 -0.821403  1.

### Answer g) 
 The ML frameworks that we will be using are:
 - LASSO REGRESSION
 - RANDOM FOREST REGRESSION

### Answer h)

In [35]:
# Now we split the dataset in train, validation and test
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_concat, y, test_size = 0.3, random_state = 42) # Split the data into 70:30 train:test

# Now we split the test dataset in validation and test
X_test, X_valid, y_test, y_valid = train_test_split(X_test, y_test, test_size = 0.5, random_state = 42) # Split the data into test and validation

# Hence we now have the train, validation and test datasets in the proportion 70:15:15

In [40]:
# Train the LASSO regression model
from sklearn.linear_model import Lasso

Lasso_model = Lasso(alpha = 0.4)
Lasso_model.fit(X_train, y_train)

Lasso(alpha=0.4)

In [41]:
# Predict the test set
y_pred = Lasso_model.predict(X_test)

# Find the mean squared error
from sklearn.metrics import mean_squared_error

MSE_lasso = mean_squared_error(y_test, y_pred)
print(MSE_lasso)

0.8038299725818752


In [44]:
# Now lets perform Cross Validation for the hyperparameter alpha
from sklearn.model_selection import cross_val_score


# Define the function to find the best alpha
def find_best_alpha(X_train, y_train):
    alphas = np.linspace(0.01, 1, 100)
    scores = []
    for alpha in alphas:
        Lasso_model = Lasso(alpha = alpha)
        scores.append(np.mean(cross_val_score(Lasso_model, X_train, y_train, cv = 5)))
    return alphas, scores


alphas, scores = find_best_alpha(X_train, y_train)

best_alpha = alphas[np.argmax(scores)]
print("The value of alpha that gives the best score is: ", best_alpha)
print("The score for the best alpha is: ", np.max(scores))

The value of alpha that gives the best score is:  0.01
The score for the best alpha is:  0.07122362200154926


In [45]:
# Evaluation of the model with best alpha using r2 score
Lasso_model = Lasso(alpha = best_alpha)
Lasso_model.fit(X_train, y_train)
y_pred = Lasso_model.predict(X_test)

Lasso_model.score(X_test, y_test)


0.06697142954874113

In [ ]:
# Random Forest Regression
from sklearn.ensemble import RandomForestRegressor

RandomForestRegressor_model = RandomForestRegressor(n_estimators = 100, max_depth = 10, random_state = 42)
RandomForestRegressor_model.fit(X_train, y_train)
